### PR TITLE
Now unnest allows bound, in and selector filters on the unnested column

### PR DESCRIPTION
**Pull request by somu-imply**
Previously even in the native query, Druid did not support filtering on the unnested column. In this PR we have introduced the following:

1. Added an allowFilter to use value matchers and advance cursors accordingly
2. Removed the allowList as we can support the same with filters

Unit tests have been added to address the following filters on the unnested column for both actual dimension and virtual columns

1. Selector filters
2. Bound filters
3. In filters

Unit tests have also been added for cases where multiple filters on unnested dimension and other dimensions on the input table are used together.

TBD: Pushing the filter on the unnested column as a filter on the dimension to be unnested on the left data source
This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by somu-imply** [comment]
The filter on the unnested dimension or dimensions(in case of an expression using multiple columns) also needs to be pushed down to the datasource to limit the number of rows the cursor goes over. This will be done in a followup PR

**Comment by somu-imply** [comment]
@317brian @techdocsmith the docs are behaving flaky with
```
> spellcheck
> mdspell --en-us --ignore-numbers --report '../docs/**/*.md' || (./script/notify-spellcheck-issues && false)

    ../docs/tutorials/tutorial-jupyter-index.md
       41 | Jupyter both try to use port `8888`, so start Jupyter on a different por 

>> 1 spelling error found in 227 files
```

The document has `port` instead of `por` so this should not be thrown. I tried running spellcheck locally and things pass

File: processing/src/main/java/org/apache/druid/segment/UnnestDimensionCursor.java
Comment (paul-rogers) [inline processing/src/main/java/org/apache/druid/segment/UnnestDimensionCursor.java:369]: As a reader, `indexedIntsForCurrentRow` doesn't seem to tell me anything. What is an "indexed int"? That is, what does it do? Are these the rows that have values for the current row? Something else? Can we come up with a name that might help us newbies?

File: processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java
Comment (paul-rogers) [inline processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java:73]: Should the "currently" and "TBD" issues be addressed before merging?

File: processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java
Comment (paul-rogers) [inline processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java:84]: Nit:

```java
final Filter forBaseFilter = filter == null || filter.getRequiredColumns().contains(outputColumnName) ? null : filter;
```

File: processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java
Comment (paul-rogers) [inline processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java:108]: Above, we _may_ have passed the filter into the "base". If we did, do we want to reapply the same filter here? Or, should the above code actually be:

```java
    final Filter forBaseFilter;
    final Filter localFilter;
    if (filter == null || filter.getRequiredColumns().contains(outputColumnName) ) {
      forBaseFilter = filter;
      localFilter = null;
    } else {
      forBaseFilter = filter;
      localFilter = null;
   }
```

Then pass `localFilter` where we're passing `filter`?

File: processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java
Comment (paul-rogers) [reply processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java:84]: This code seems to assume a single filter. In the example, `unnest-d2 < 10`. But, what if my condition is `unnest-d2 < 10 AND unnest_d2 > 5`. Will this code work? And if only the `unnest-d2` appears here, do we need to check the output column name?

To help understand this, it would be good to give a bit more of the query. What, exactly, is `d2` here?

File: processing/src/test/java/org/apache/druid/query/groupby/UnnestGroupByQueryRunnerTest.java
Comment (paul-rogers) [inline processing/src/test/java/org/apache/druid/query/groupby/UnnestGroupByQueryRunnerTest.java:699]: Sorry, what is a `PLACEMENTISH`?

File: processing/src/test/java/org/apache/druid/segment/UnnestColumnValueSelectorCursorTest.java
Comment (paul-rogers) [inline processing/src/test/java/org/apache/druid/segment/UnnestColumnValueSelectorCursorTest.java:390]: Should these tests be retained, but updated to whatever the new format is? Presumably these addressed some earlier concern, and we should ensure that they still work. Or, are there new tests that cover the same cases that the deleted tests cover?

File: processing/src/test/java/org/apache/druid/segment/UnnestColumnValueSelectorCursorTest.java
Comment (paul-rogers) [inline processing/src/test/java/org/apache/druid/segment/UnnestColumnValueSelectorCursorTest.java:54]: Do these tests a wide range of predicates (filters) on the unnested value? I worry that the code above handles only simple cases: `unnested_value op constant`. Do they also handle complex cases such as `fn(unnested_value) > some_other_col`? `fn(unnested_value) > 10 AND unnested_value * 3 > 12`? Would be good to have tests that exercise these non-trivial cases.

**Review by paul-rogers — CHANGES_REQUESTED** [review]

File: processing/src/main/java/org/apache/druid/segment/UnnestColumnValueSelectorCursor.java
Comment (somu-imply) [reply processing/src/main/java/org/apache/druid/segment/UnnestColumnValueSelectorCursor.java:264]: Thanks this should be 
```
while (true) {
      advanceAndUpdate();
      boolean match = valueMatcher.matches();
      if (match || baseCursor.isDone()) {
        return;
      }
    }
```
The match changes after the update so the call should be below. I'll add it

File: processing/src/main/java/org/apache/druid/segment/UnnestDimensionCursor.java
Comment (somu-imply) [reply processing/src/main/java/org/apache/druid/segment/UnnestDimensionCursor.java:109]: I did not understand this comment, can you please elaborate. On another note I'll remove the bitset it is not needed anymore

File: processing/src/main/java/org/apache/druid/segment/UnnestDimensionCursor.java
Comment (somu-imply) [reply processing/src/main/java/org/apache/druid/segment/UnnestDimensionCursor.java:208]: Valid point, we can only estimate it correctly for In filters (no in the in filter) and selector filter (should always be 1). This can get complicated for a filter such as d3='a' AND d3='b' where the cardinality will be equal to the size of the filters. I think Calcite would change that to d3 IN (a,b) so we should only handle the 2 cases

File: processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java
Comment (somu-imply) [reply processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java:84]: I'll add more comment. d2 is the column to be unnested. What I mean is that all filters on columns that exist in the datasource are pushed on to the data source. Only the filters on the unnested column appears here

File: processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java
Comment (somu-imply) [reply processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java:108]: The filter that appears here will be on the unnested column only. For example if you are unnesting dim3 to ud3, any filter on ud3 will appear here. The filter on ud3 cannot be pushed into base but can only be sent to the unnest cursor as an allow filter. There is scope of optimization here to rewrite the same filter on dim3 and push it to the base

File: processing/src/test/java/org/apache/druid/query/groupby/UnnestGroupByQueryRunnerTest.java
Comment (somu-imply) [reply processing/src/test/java/org/apache/druid/query/groupby/UnnestGroupByQueryRunnerTest.java:699]: It is a dimension in our test setup that is of the format `placement,<some_string>`

File: processing/src/test/java/org/apache/druid/segment/UnnestColumnValueSelectorCursorTest.java
Comment (somu-imply) [reply processing/src/test/java/org/apache/druid/segment/UnnestColumnValueSelectorCursorTest.java:390]: There are new tests that cover these. They are added in the CalciteArraysQueryTest

**Comment by somu-imply** [comment]
The one failing test here will be resolved once https://github.com/apache/druid/pull/13860 gets merged

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteArraysQueryTest.java
Comment (clintropolis) [inline sql/src/test/java/org/apache/druid/sql/calcite/CalciteArraysQueryTest.java:3403]: please add a test where there are filters on other columns too (a mix of filter on unnested column and filter on some other column)

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java
Comment (clintropolis) [inline sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:240]: can you not have both a where filter an an unnest filter? is an unnest filter basically a where filter that got picked up somewhere else?

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteArraysQueryTest.java
Comment (clintropolis) [reply sql/src/test/java/org/apache/druid/sql/calcite/CalciteArraysQueryTest.java:3371]: oh, i got confused from all the repetitive patterns in this query, d3 isn't dim3, another instance of me not being able to read :sob:

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteArraysQueryTest.java
Comment (somu-imply) [reply sql/src/test/java/org/apache/druid/sql/calcite/CalciteArraysQueryTest.java:3403]: There is 1 already `SELECT d12 FROM druid.numfoo, UNNEST(ARRAY[m1, m2]) as unnested (d12) where d12 IN ('1','2') AND m1 < 10`

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java
Comment (somu-imply) [reply sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:240]: Yes an unnest filter is same as a where or having that gets picked up somewhere else

**Comment by somu-imply** [comment]
>It seems like you could potentially decompose the filter to partially push down anything that is is not filtering on the output column name into the base cursor so that the query can still use indexes and such, since otherwise the implication here is that if any filter is on the unnest column then it means we aren't going to use any indexes since we aren't pushing filter down to the base cursor.

Calcite is smart enough to push the filter on an existing dimension to the underlying datasource. For example a query like
```with t as (select * from druid.numfoo,UNNEST(MV_TO_ARRAY(dim3)) as foo(d3))
select d3 from t where d3 >1 AND dim2 IN ('a','b')```

Calcite plans it as
```
157:LogicalProject(d3=[$17])
  155:LogicalCorrelate(subset=[rel#156:Subset#7.NONE.[]], correlation=[$cor0], joinType=[inner], requiredColumns=[{3}])
    144:LogicalFilter(subset=[rel#145:Subset#1.NONE.[]], condition=[OR(=($2, 'a'), =($2, 'b'))])
      9:LogicalTableScan(subset=[rel#143:Subset#0.NONE.[]], table=[[druid, numfoo]])
    151:LogicalFilter(subset=[rel#152:Subset#5.NONE.[]], condition=[>(CAST($0):INTEGER, 1)])
      149:Uncollect(subset=[rel#150:Subset#4.NONE.[]])
        147:LogicalProject(subset=[rel#148:Subset#3.NONE.[]], EXPR$0=[MV_TO_ARRAY($cor0.dim3)])
          10:LogicalValues(subset=[rel#146:Subset#2.NONE.[0]], tuples=[[{ 0 }]])
```

File: processing/src/main/java/org/apache/druid/segment/UnnestDimensionCursor.java
Comment (imply-cheddar) [reply processing/src/main/java/org/apache/druid/segment/UnnestDimensionCursor.java:208]: We cannot eliminate the bitset here.  You need to be producing the bitset by applying the value matcher to the dictionary of values.

File: processing/src/main/java/org/apache/druid/segment/UnnestDimensionCursor.java
Comment (imply-cheddar) [reply processing/src/main/java/org/apache/druid/segment/UnnestDimensionCursor.java:361]: IndexedInt is a class in the Druid code base.  It's logically equivalent to an `int[]`.  The values are references to the dictionary.  There is some javadoc that attempts to explain these things on interfaces like `DimensionSelector` and `DimensionDictionarySelector`

File: processing/src/main/java/org/apache/druid/query/UnnestDataSource.java
Comment (imply-cheddar) [inline processing/src/main/java/org/apache/druid/query/UnnestDataSource.java:56]: I had expected you to change the allowList to a Filter but keep it on this object.  It looks like you are assuming that the filter passed into the call to make the cursor is good enough.  In the "best" case, you are planning things to attach the filter on the unnested thing to the other filters being included, in which case, as clint mentions, the entire filter is moving to a value matcher with absolutely nothing being pushed down.  In the worst case, some of the filters are being lost in planning and incorrect results can occur.  I haven't validated which one of these is likely occuring yet.

Given that Calcite very clearly puts the filter on the UNCOLLECT, you should be able to just attach a Filter object here instead of putting it on the query as a whole.  The logic inside of this should remain more or less the same as what it was before with the allow list, just, you are applying a value predicate to the dictionaries instead of building a bitset.

The filter used to make the cursor should continue to be passed down (though it should also have the filter that is attached here adjusted and pushed down as well).

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteArraysQueryTest.java
Comment (imply-cheddar) [inline sql/src/test/java/org/apache/druid/sql/calcite/CalciteArraysQueryTest.java:3522]: So, from looking at how things are planned, I thought that we were throwing away the actual where clause.  Instead, this test showed me what is actually happening: when we have both filters, the unnest is being planned on top of a scan query which carries the original filter.

This is a bad plan, it is going to needlessly force things to run on the Broker instead of being pushed down.  The unnest should be pushed down onto the actual TableDataSource and the native query here should be a singular scan on top of an unnest data source of a table reference.

**Review by imply-cheddar — CHANGES_REQUESTED** [review]
The way that this is planning is going to create very sub-optimal runs and limitations on query running that shouldn't exist.

The code strategy should've been to replace the allow list with a Filter object and then work with that.  We cannot merge this code as is and must change it around to replacing the allow list with a Filter object and then applying that.

**Comment by somu-imply** [comment]
The following things have been addressed:

1. Filter moved inside unnest data source
2. A transformation is introduced that creates the partial query on the correlate from the partial query on the left
3. We only create a query data source if there is a LIMIT as the limit needs to be honored before unnest
4. Updated test cases.

Things to be done:

1. Add more tests
2. Modify getValueCardinality for the dimension cursor
3. Transform the filter on the unnested output column to filters on the left and pass them to the base cursor

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/PartialDruidQuery.java
Comment (somu-imply) [inline sql/src/main/java/org/apache/druid/sql/calcite/rel/PartialDruidQuery.java:87]: This needs to be removed

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidCorrelateUnnestRel.java
Comment (imply-cheddar) [inline sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidCorrelateUnnestRel.java:276]: This feels like we are throwing away the left-hand side and instead trying to build a query from the unnest part?  Taht seems wrong to me, we should be using the unnest on the RHS to augment the query from the LHS (attach to the datasource).  the LHS is the actual working query, the RHS is just the addition of the unnest portion.

**Review by imply-cheddar — COMMENTED** [review]
Leaving some comments here.  As I did the review, I *think* that you are likely still in the process of adjusting things as some of the stuff that we've discussed directly isn't yet reflected here.

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidCorrelateUnnestRel.java
Comment (somu-imply) [reply sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidCorrelateUnnestRel.java:148]: Yes we have a test case, the query is like
`SELECT d3 FROM (select * from druid.numfoo where dim2 IN ('a','b','ab','abc') LIMIT 2), UNNEST(MV_TO_ARRAY(dim3)) as unnested (d3)`

Here the limit from the left cannot be applied outside

File: processing/src/main/java/org/apache/druid/segment/UnnestDimensionCursor.java
Comment (somu-imply) [reply processing/src/main/java/org/apache/druid/segment/UnnestDimensionCursor.java:109]: Yes I am workin on the logic to add the bitset back.

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidCorrelateUnnestRel.java
Comment (somu-imply) [reply sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidCorrelateUnnestRel.java:276]: The DruidCorrelateUnnestRule already put the select project on top of the correlateRel by adding all the things in right to the left. Also the group by, order by on the outer query are already present on the correlateRel. So we create the query by adding the select project (the one which is already projected on top with left + right), the filters from the left + aggregates , sorts if present on the top level

**Comment by gianm** [comment]
@somu-imply could you please have a look at https://github.com/apache/druid/pull/13892? That has some changes that I started doing as part of looking into upgrading Calcite. The upgrade borked a bunch of the existing UNNEST tests and exposed some issues with how the planning works, which led to invalid plans. The changes in the #13892 should fix up the planning. IMO the logic is also cleaner.

**Pull request closed without merging**
